File: 백준/Gold/9663. N－Queen/N－Queen.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	
	static int n;
	static boolean[] c;
	static boolean[] bslash;
	static boolean[] slash;
	static int cnt;
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		n = Integer.parseInt(br.readLine());
		cnt = 0;
		c = new boolean[n];
		bslash = new boolean[2*(n-1)+1];
		slash = new boolean[2*(n-1)+1];
		
		
		dfs(0);
		System.out.println(cnt);
		
		
	}
	private static void dfs(int k) {
		if (k==n) {
			cnt++;
			return;
		}
		
		for(int i=0; i<n; i++) {
			if(!c[i] && !bslash[k+i] && !slash[(n-1+k-i)]) {
				c[i] = true;
				bslash[i+k] = true;
				slash[(n-1)+k-i] = true;
				dfs(k+1);
				slash[(n-1)+k-i] = false;
				bslash[i+k] = false;
				c[i] =false;
				
			}
		}
		
	}

}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline

n = int(input())

c = [0]*n
bslash = [0]*(2*(n-1)+1)
slash = [0]*(2*(n-1)+1)
cnt = 0

def dfs(k):
    global n, cnt
    if k == n:
        cnt+=1
        return 
    
    for i in range(n):
        if not c[i] and not bslash[k+i] and not slash[(n-1)+k-i]:
            c[i] = 1
            bslash[i+k] =1
            slash[(n-1)+k-i] = 1
            dfs(k+1)
            slash[(n-1)+k-i] = 0
            bslash[i+k] =0
            c[i] = 0

dfs(0)
print(cnt)

            
